### PR TITLE
[#110][#111] List operation handles large collections

### PR DESCRIPTION
**Pull request by korydraughn**
This needs to be tested.

### CODE REVIEW

**Comment by trel** [comment]
seems good - ready to # if we're tested/confirmed.      what changed in the force push?

**Comment by korydraughn** [comment]
Working on and running test right now. The force push involved a change to the pom.xml file. I needed to revert part of the file so that the build succeeds.

**Comment by trel** [comment]
got it - thanks.

**Comment by korydraughn** [comment]
I've added a BATS test that verifies that the list operation produces the correct number of entries for a *large* collection (in this case 3000 files). The test passed.

We'll have to look into whether this PR resolves #110. We have need to use apache httpd to simulate that issue.

**Comment by trel** [comment]
great - yes, let's see if sanger may want to test this branch themselves, too.     if you're not seeing hitches in the listings anymore, i'm fine to get it #'d and merged.   can always leave the issue open until they confirm as fixed.

**Comment by trel** [comment]
@kript @bh9 please eyeball this when you get a chance

**Comment by michael-conway** [comment]
Sounds great, let me know how it's going. I'm interested to see how the
removed ObjStat will improve listing. There remain some interior changes to
the list all method I can do under the covers too, mostly in how it's
paging. We can talk about that as a separate issue once we get the listings
working properly.

**Comment by kript** [comment]
This sounds exiting! However I shall defer to my esteemed colleague @ac55-sanger who is leading the charge here....

**Comment by ac55-sanger** [comment]
Sure, sounds good, happy to test.
Will update once done. Thanks.

**Comment by ac55-sanger** [comment]
Hi,

docker build is failing on korydraughn:110 with sleepycat dependency issue:

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] nfs4j-irodsvfs ..................................... SUCCESS [  1.144 s]
[INFO] nfsrods ............................................ FAILURE [03:58 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 04:03 min
[INFO] Finished at: 2021-01-26T14:08:52+00:00
[INFO] Final Memory: 20M/714M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project nfsrods: Could not resolve dependencies for project org.irods.jargon:nfsrods:jar:1.0.1: Could not find artifact com.sleepycat:je:jar:7.3.7 in dcache-snapshots (https://download.dcache.org/nexus/content/repositories/releases) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :nfsrods
The command '/bin/sh -c cd irods_client_nfsrods &&     git checkout ${_sha} &&     mvn clean install -Dmaven.test.skip=true' returned a non-zero code: 1
```

I tried to add it in pom.xml and re-build but didn't work. 
Could you please have a look and let me know?
Thanks!

**Comment by korydraughn** [comment]
Will look into this.

**Comment by korydraughn** [comment]
@ac55-sanger Mike and I have resolved the build issue. Please give this another shot when you get a chance.

**Comment by ac55-sanger** [comment]
Sure, thanks.
Will update.

**Comment by ac55-sanger** [comment]
@korydraughn 
This is still failing to build with same 
'''
Could not resolve dependencies for project org.irods.jargon:nfsrods:jar:1.0.1: Could not find artifact com.sleepycat:je:jar:7.3.7 in dcache-snapshots
'''
dependency issue.

**Comment by korydraughn** [comment]
Your docker build command is actually trying to build the master branch. You need to point that command at this branch. To do that, run the following:
```bash
$ docker build -t nfsrods --build-arg='_github_account=korydraughn' --build-arg='_sha=110' 
```

**Comment by ac55-sanger** [comment]
Ah, my bad. Thanks, it built successfully. 
I will test it next week and let you guys know.

**Comment by ac55-sanger** [comment]
@korydraughn 
I tested this build and here is my report:

1. "ls" on the directory containing 65K+ files took more than 24 hours and yet resulted in wrong data and count. 
```
$ ls -l | wc -l
.
ls: cannot access 'DDD_MAIN5249029.cram': No such file or directory
ls: cannot access 'DDD_MAIN5249029.cram.crai': No such file or directory
ls: cannot access 'DDD_MAIN5249030.cram': No such file or directory
ls: cannot access 'DDD_MAIN5249030.cram.crai': No such file or directory
1631
```
whereas all these files exist on irods and I can list these using "ils" command.

2. This build also broke data mapping and random data is being mapped to a collection. 
```
$ ls -l <mounted_path>/20140918/
ls: cannot access '<mounted_path>/20140918/cram': No such file or directory
total 0
d????????? ? ? ? ?            ? cram

$ ls -l <mounted_path>/20140918/ | wc -l
ls: cannot access '/mnt/humgen/projects/ddd/20140918/cram': No such file or directory
2

=====
$ ils <irods_path>/20140918/
<irods_path>/20140918:
  10:1-135534747.vcf.gz
  10:1-135534747.vcf.gz.tbi
  11:1-135006516.vcf.gz
  11:1-135006516.vcf.gz.tbi
.
.
.
$ ils <irods_path>/20140918/ | wc -l
49

```
NFSRODs mounted directory 1) doesn't list the data properly 2) lists some random data (eg. "cram" directory in this case) which doesn't exist in irods.

This mapping issue is not seen in the previous release build (tested it again).

Thanks

**Comment by michael-conway** [comment]
Kory let's get together on this.

**Comment by korydraughn** [comment]
We'll continue to look into this.

**Comment by korydraughn** [comment]
@ac55-sanger When you did your testing, did you make sure to unmount nfsrods before remounting it?

I know that weird things happen when the nfsrods server is bounced without remounting it.

**Comment by michael-conway** [comment]
fwiw Kory, Deep and I are setting up a test platform with 65K+ for our own
testing and for profiling/optimizatoin

**Comment by ac55-sanger** [comment]
> @ac55-sanger When you did your testing, did you make sure to unmount nfsrods before remounting it?
> 
> I know that weird things happen when the nfsrods server is bounced without remounting it.

Yes @korydraughn I remember unmounting the old one before mounting a server with the new build. 
But I can re-check and let you know by the end of today.

**Comment by korydraughn** [comment]
@ac55-sanger Can you give this PR another try?

I believe the issues have been resolved. `wc -l` outputs the expected numbers. I was able to list 65k files under 5 minutes with the latest changes. Setting the log level to **warn** also improved performance.

**Comment by korydraughn** [comment]
Using seconds is motivated by how they are used in the implementation (and they avoid a lot of zeros). Those options aren't useful when the values are less than one second. I figured it would be easier to think in seconds rather than in milliseconds as well.

I agree that mixing them could be lead to issues. Figuring out very good defaults could help to drive whether they should be defined in seconds or milliseconds as well.

It would interesting to know whether users actually tweak these values at all.

**Comment by ac55-sanger** [comment]
@korydraughn 
Built the docker image from the branch (110). But while mounting NFSRODs, it's getting hanged with connection refused error:
```
mount.nfs: portmap query failed: RPC: Unable to receive - Connection refused
```

docker logs command shows below warnings:
```
2021-04-08 11:35:57.296 DEBUG Thread-1 [IRODSIdMapper] - map - mapping inode number to path [1 => /humgen] ...
2021-04-08 11:35:57.296 DEBUG Thread-1 [IRODSIdMapper] - map - previously mapped path [null]
2021-04-08 11:35:57.297 DEBUG Thread-1 [IRODSIdMapper] - establishRoot - Mapping successful.
2021-04-08 11:36:05.472 WARN  Thread-1 [OncRpcSvc] - Failed to register at portmap: portmap service not available
2021-04-08 11:36:05.484 INFO  Thread-1 [ServerMain] - main - Press any key to shutdown.
2021-04-08 11:36:05.484 INFO  Thread-1 [ServerMain] - main - Shutting down ...
2021-04-08 11:36:09.501 INFO  Thread-1 [ServerMain] - main - done.
```

Have you come across this?

Thanks,
Ashwini

**Comment by korydraughn** [comment]
The warning in the log file is common. I haven't figured out what causes it, but that doesn't stop NFSRODS from working properly.

The `Connection refused` message is more concerning. You're receiving that when running `mount`? That  is new to me. I'll see if I can reproduce it.

**Comment by ac55-sanger** [comment]
Yes, I receive that error when I run the mount command. 
Once I revert back to the old build, the exact mount command succeeds.

**Comment by korydraughn** [comment]
That is very strange. Can you provide the mount command you are running?

**Comment by korydraughn** [comment]
The issue appears to stem from docker and this branch. I'll investigate and let you know when it is ready.

You can always build and run NFSRODS outside of a container. That is how I run and test it. Hopefully, removing docker from the equation will allow you to test it if you choose to go that route. All you would need to do is install the dependencies, build the jar files, and run it. All of those steps are in the Dockerfile.

**Comment by ac55-sanger** [comment]
We have ansible setup for mounting, so that was failing. 
Manually, as there is an entry in /etc/fstab, 
```
irods-hum-nfsrods01.internal.sanger.ac.uk:/ /mnt/humgen nfs vers=4.1,tcp,rsize=1048576,wsize=1048576,timeo=600,hard,intr,noatime,ro 0 0
```
I could just do `mount -a` and it'd mount, but that fails with a timeout error. 

And then I also tested below command:
```
mount -v -o sec=sys,port=2049 irods-hum-nfsrods01.internal.sanger.ac.uk:/ /mnt/humgen
```
This also resulted in the same timeout error.

**Comment by ac55-sanger** [comment]
> The issue appears to stem from docker and this branch. I'll investigate and let you know when it is ready.
> 
> You can always build and run NFSRODS outside of a container. That is how I run and test it. Hopefully, removing docker from the equation will allow you to test it if you choose to go that route. All you would need to do is install the dependencies, build the jar files, and run it. All of those steps are in the Dockerfile.

I would like to avoid this approach as I'm directly testing on a production machine that is ansiblised, and would avoid the dependencies installed on the machine directly.

**Comment by korydraughn** [comment]
Just an FYI, the NFSRODS server is not standing up properly when launched in a docker container. That explains the connection refused error message. I hit the issue you described as soon as I launched it in a container.

**Comment by korydraughn** [comment]
> > The issue appears to stem from docker and this branch. I'll investigate and let you know when it is ready.
> > You can always build and run NFSRODS outside of a container. That is how I run and test it. Hopefully, removing docker from the equation will allow you to test it if you choose to go that route. All you would need to do is install the dependencies, build the jar files, and run it. All of those steps are in the Dockerfile.
> 
> I would like to avoid this approach as I'm directly testing on a production machine that is ansiblised, and would avoid the dependencies installed on the machine directly.

No problem. I will work to figure out why NFSRODS is not launching correctly in the container and let you know what I find.

**Comment by ac55-sanger** [comment]
Sure, thanks.

**Comment by korydraughn** [comment]
@ac55-sanger Please pull the latest changes and try running it again. I think we got it.

Because I changed how the server waits for the shutdown signal, you now have to pass the `-i` flag to `docker run` in addition to the other flags. That will provide the server with an interactive console even in detached mode. This is required because the server wants to read from `stdin`.

**Comment by ac55-sanger** [comment]
@korydraughn 
This doesn't look good either. I have tested it twice to make sure I'm not missing anything.

I'm getting mixed results. 

1. Basic directory browsing works fine and listing of directory contents works as expected if the number of files is less. (Always works)
2. Listing contents of a directory with 900 files, 
The first time it'd get stuck mostly, and some times it'd work okayish. 
The second (and further) time when it's served from cache, it'd work. 
(quite unreliable) 
3. Listing contents of a directory with 65K files: It was stuck for ~20 hours with no results, I had to cancel that later. 
(Never worked).

Thanks,
Ashwini

**Comment by korydraughn** [comment]
That is surprising. Can you verify that the SHA for your build matches the one for this PR?

**Comment by ac55-sanger** [comment]
Sure,

```
~# docker run --rm ac55/nfsrods-patch:3.0 sha
Build Time    => 2021-04-14T11:29:39+0000
Build Version => 1.0.1
Build SHA     => 35e278cd3b89f4809f8859fe7e85197baccf24f9
```

**Comment by ac55-sanger** [comment]
It just gave up after ~38 hours. :(

```
ls: reading directory '<path_to_mounted_directory>': Remote I/O error
total 0

real    2233m20.581s
user    0m0.000s
sys     0m0.003s
```

**Comment by trel** [comment]
We narrowed the scope of @ac55-sanger's slow listings to an iRODS specific query that is not valid syntax for Oracle.  We are moving ahead with the rest of these edits and will tackle the Oracle syntax issue separately.

**Comment by korydraughn** [comment]
@michael-conway Can you make a new snapshot of jargon (tip of master) available?

**Comment by korydraughn** [comment]
@ac55-sanger Please try using NFSRODS again. You'll need to rebuild the docker image.

You'll need to make some changes before running NFSRODS.
- Add `"using_oracle_database": true` to the `"nfs_server"` section of the NFSRODS config file.
- Replace the following specific queries in iRODS with ones that work with Oracle.
  - `ilsLACollections`
  - `ilsLADataObjects`

Below are the new specific queries. Please have Simon take a look at these. You're free to adjust these.

#### ilsLACollections
```sql
SELECT * FROM (
  SELECT rownum as limit_rn, c.parent_coll_name, c.coll_name, c.create_ts, c.modify_ts,
         c.coll_id, c.coll_owner_name, c.coll_owner_zone, c.coll_type, u.user_name, u.zone_name,
         a.access_type_id, u.user_id
  FROM R_COLL_MAIN c
  JOIN R_OBJT_ACCESS a ON c.coll_id = a.object_id
  JOIN R_USER_MAIN u ON a.user_id = u.user_id
  WHERE c.parent_coll_name = ?
  ORDER BY c.coll_name, u.user_name, a.access_type_id, c.parent_coll_name, c.create_ts, c.modify_ts,
           c.coll_id, c.coll_owner_name, c.coll_owner_zone, c.coll_type, u.zone_name, u.user_id DESC
) WHERE limit_rn > ? AND limit_rn <= ?
```

#### ilsLADataObjects
```sql
SELECT * FROM (
  SELECT rownum as limit_rn, s.coll_name, s.data_name, s.create_ts, s.modify_ts, s.data_id,
         s.data_size, s.data_repl_num, s.data_owner_name, s.data_owner_zone, u.user_name,
         u.user_id, a.access_type_id, u.user_type_name, u.zone_name
  FROM (
      SELECT c.coll_name, d.data_name, d.create_ts, d.modify_ts, d.data_id, d.data_repl_num,
             d.data_size, d.data_owner_name, d.data_owner_zone
      FROM R_COLL_MAIN c
      JOIN R_DATA_MAIN d ON c.coll_id = d.coll_id
      WHERE c.coll_name = ?
  ) s
  JOIN R_OBJT_ACCESS a ON s.data_id = a.object_id
  JOIN R_USER_MAIN u ON a.user_id = u.user_id
  ORDER BY s.coll_name, s.data_name, u.user_name, a.access_type_id, s.create_ts, s.modify_ts,
           s.data_id, s.data_size, s.data_repl_num, s.data_owner_name, s.data_owner_zone,
           u.user_id, u.user_type_name, u.zone_name DESC
) WHERE limit_rn > ? and limit_rn <= ?
```

**Comment by korydraughn** [comment]
@ac55-sanger Please hold on replacing the existing specific queries.

**Comment by ac55-sanger** [comment]
Sure. Thankfully I haven't looked into these yet.

**Comment by korydraughn** [comment]
@ac55-sanger How long does it take `ils -A` to run against a fairly large collection on your system?

I'm trying to determine if anything breaks by changing those queries. I feel anything attempting to invoke the existing ones will result in the poor performance we see in NFSRODS.

**Comment by ac55-sanger** [comment]
Here is the time taken to list a directory having 65K files using ils command -

```
real	3m31.221s
user	0m15.197s
sys	0m4.812s
```

**Comment by korydraughn** [comment]
@ac55-sanger You can proceed with replacing those specific queries. They were added to improve Jargon's performance around large data sets.

**Comment by ac55-sanger** [comment]
Sure. 
Should I go ahead with the queries shared last week or the ones mentioned in 

```
See https://github.com/irods/irods-legacy/blob/ff4eaa47a34f1bb5990d5560f825975c26bab118/iRODS/server/icat/patches/patch3.2to3.3.sh
```

?

**Comment by korydraughn** [comment]
Don't use the ones from irods-legacy. Use the replacements mentioned here: https://github.com/irods/irods_client_nfsrods/pull/114#issuecomment-840821127

**Comment by ac55-sanger** [comment]
Sure, thanks. 
Will update on how it goes.

File: Dockerfile [hunk 1]
```diff
@@ -1,10 +1,9 @@
 FROM ubuntu:16.04
-ARG sssd=false
```
Comment (trel) [inline Dockerfile:2]: this is... important for the boulder use case?

Comment (korydraughn) [reply Dockerfile:2]: I believe that isn't used anywhere else in the Dockerfile which is why i removed it.

Comment (korydraughn) [reply Dockerfile]: restored ARG.

**Pull request merged**